content/blog/insecure-software-development/index.md

```diff
@@ -0,0 +1,110 @@
+
+### Encapsulate complexity, abstract
+
+The idea of this approach is to _hide complexity_ so developers don't need to understand all the complexities of a certain domain in order to work on a codebase. This is usually related to classes or modules, where each one is designed to be relatively independent of each other. Developers can work in a module without concerning themselves with another. Abstractions can go wrong in two ways though:
+
+1. Expose information that is not important (leaky abstraction)
+2. Omit details that are important (obscure abstractions or false abstractions)
+
+Leaky abstractions are painful to handle from a developer perspective, but don't affect security as much as false abstractions. Dependency confusion can be explained by false abstractions. By not exposing where dependencies were being resolved from, which is an important detail on this domain, package managers created a false abstraction that allowed attackers to exploit package resolution. Developers didn't have the right information in order to use the abstraction correctly.
+
+## Security
+
+From a security perspective _designing for simplicity_ is the best way to achieve secure code. And simplicity here doesn't only mean less features or code, it means making things more obvious. You can have a functionality that requires hundreds of lines of codes that is simpler than a similar one with half of that code.
+
+Let's use the old Javascript or PHP type juggling as scenarios where code simplicity was not followed:
+
+```js
+"0e1234" == 0
+true
+```
+
+or
+
+```php
+var_dump('0xCAFE' == 51966);
+bool(true)
+```
+
+_This is fixed in newer versions of PHP_
+
+It's not obvious that type casting will happen during comparison and actually make these comparisons return true. Being forced to remember special cases of your daily functions increases cognitive load and leads to bugs. Unsurprisingly, we had a range of vulnerabilities occurring because of this implicit behaviour. Plenty of other programming languages and libraries have implicit behaviour. From URL parsers causing _URL confusion_ to user management software not employing the principle of least privilege by default.
+
+So what can we do to reduce hidden complexity in our own systems? This is a really hard question, but code reviews always comes to my mind as a place where software development and security can work together to generate a better outcome.
+
+## Code Reviews
+
+Code reviews are a great way to identify complex or implicit behaviour. It doesn't matter if you are a junior, senior or the ultimate developer, if someone points out that they are having a hard time understanding your code then your code is complex.
+
+As a code reviewer you have the opportunity to:
```

Comment:
What a great checklist for code reviews. 👏🏻 👌🏻 
![image](https://user-images.githubusercontent.com/26663077/151421075-bfe1e2e5-f1ad-48ae-a37f-95a3257f9791.png)